### PR TITLE
Fixes #9403 and #10803: missing flattening of nested applications in notations

### LINKED ISSUE #9403: Notation stops printing when defined using other notation
#### Version

8.9 (branch)

#### Operating system

Linux

#### Description of the problem

The following example
```coq
From Coq Require Export Utf8.

Local Set Universe Polymorphism.

Module tele.
(** Telescopes *)
Inductive tele : Type :=
  | TeleO : tele
  | TeleS {X} (binder : X → tele) : tele.

Arguments TeleS {_} _.

(** The telescope version of Coq's function type *)
Fixpoint tele_fun (TT : tele) (T : Type) : Type :=
  match TT with
  | TeleO => T
  | TeleS b => ∀ x, tele_fun (b x) T
  end.

Notation "TT -t> A" :=
  (tele_fun TT A) (at level 99, A at level 200, right associativity).

(** An eliminator for elements of [tele_fun].
    We use a [fix] because, for some reason, that makes stuff print nicer
    in the proofs in iris:bi/lib/telescopes.v *)
Definition tele_fold {X Y} {TT : tele} (step : ∀ {A : Type}, (A → Y) → Y) (base : X → Y)
  : (TT -t> X) → Y :=
  (fix rec {TT} : (TT -t> X) → Y :=
     match TT as TT return (TT -t> X) → Y with
     | TeleO => λ x : X, base x
     | TeleS b => λ f, step (λ x, rec (f x))
     end) TT.
Arguments tele_fold {_ _ !_} _ _ _ /.

(** A sigma-like type for an "element" of a telescope, i.e. the data it
  takes to get a [T] from a [TT -t> T]. *)
Inductive tele_arg : tele → Type :=
| TargO : tele_arg TeleO
(* the [x] is the only relevant data here *)
| TargS {X} {binder} (x : X) : tele_arg (binder x) → tele_arg (TeleS binder).

Definition tele_app {TT : tele} {T} (f : TT -t> T) : tele_arg TT → T :=
  λ a, (fix rec {TT} (a : tele_arg TT) : (TT -t> T) → T :=
     match a in tele_arg TT return (TT -t> T) → T with
     | TargO => λ t : T, t
     | TargS x a => λ f, rec a (f x)
     end) TT a f.
Arguments tele_app {!_ _} _ !_ /.

Coercion tele_arg : tele >-> Sortclass.
Local Coercion tele_app : tele_fun >-> Funclass.

(** Operate below [tele_fun]s with argument telescope [TT]. *)
Fixpoint tele_bind {U} {TT : tele} : (TT → U) → TT -t> U :=
  match TT as TT return (TT → U) → TT -t> U with
  | TeleO => λ F, F TargO
  | @TeleS X b => λ (F : TeleS b → U) (x : X), (* b x -t> U *)
                  tele_bind (λ a, F (TargS x a))
  end.
Arguments tele_bind {_ !_} _ /.

(** Notation-compatible telescope mapping *)
(* This adds (tele_app ∘ tele_bind), which is an identity function, around every
   binder so that, after simplifying, this matches the way we typically write
   notations involving telescopes. *)
Notation "t $ r" := (t r)
  (at level 65, right associativity, only parsing).
Notation "'λ..' x .. y , e" :=
  (tele_app $ tele_bind (λ x, .. (tele_app $ tele_bind (λ y, e)) .. ))
  (at level 200, x binder, y binder, right associativity,
   format "'[  ' 'λ..'  x  ..  y ']' ,  e").

(** Telescopic quantifiers *)
Definition texist {TT : tele} (Ψ : TT → Prop) : Prop :=
  tele_fold ex (λ x, x) (tele_bind Ψ).
Arguments texist {!_} _ /.

Notation "'∃..' x .. y , P" := (texist (λ x, .. (texist (λ y, P)) .. ))
  (at level 200, x binder, y binder, right associativity,
  format "∃..  x  ..  y ,  P").
End tele.
Import tele.

(* This is like Iris' accessors, but in Prop.  Just to play with telescopes. *)
Definition accessor {X : tele} (α β γ : X → Prop) : Prop :=
  ∃.. x, α x ∧ (β x → γ x).

(* Working with abstract telescopes. *)
Section tests.
Context {X : tele}.
Implicit Types α β γ : X → Prop.

Lemma acc_mono_disj α β γ1 γ2 :
  accessor α β γ1 → accessor α β (λ.. x, γ1 x ∨ γ2 x).
Proof.
  Show.
```
prints
```
  accessor α β γ1 → accessor α β (tele_app (tele_bind (λ x : X, γ1 x ∨ γ2 x)))
```
As you can see, the notation `λ..` is not used. The funny thing is, if I change this to define that notation as
```coq
Notation "'λ..' x .. y , e" :=
  (tele_app (tele_bind (λ x, .. (tele_app (tele_bind (λ y, e))) .. )))
  (at level 200, x binder, y binder, right associativity,
   format "'[  ' 'λ..'  x  ..  y ']' ,  e").
```
then it works. All I did is unfold the `$` notation. That's a parsing-only notation, so it definitely should not have any effect on printing -- and yet somehow, using or not using it does make the difference here.

### LINKED ISSUE #10803: Nested notations break bound scopes
#### Description of the problem

The last of these should not fail, I think.

```coq
Inductive Foo := foo.
Delimit Scope foo_scope with foo.
Bind Scope foo_scope with Foo.
Notation "'!'" := foo : foo_scope.
Definition of_foo {x : nat} {y : nat} (f : Foo) := f.
Notation a := (@of_foo O).
Notation b := (@a).
Check a !.
Check @a O !.
Check @b O.
Check @b O !. (* Error: Unknown interpretation for notation "!". *)
```

See Also #8126

cc @herbelin 

#### Coq Version

8.9.0, also master post v8.10 (59079a232d2157c0c4bea4cb1a3cd68c9410e880)

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

A trivial bug with a trivial fix.

Fixes #9403
Fixes #10803

- [X] Added / updated test-suite
- [x] Entry added in the changelog

### CODE REVIEW

**Comment by herbelin** [comment]
@JasonGross, may I get your feedback on this PR?

The failure in rewriter is due to the following situation:
```
(* arg_types_of is an application of 7 arguments to ident.arg_types, with the 8th argument being a maximal implicit *)
Notation arg_types_of p := (@ident.arg_types _ _ (@raw_ident _ _ p) (@raw_ident_infos_of _ _ p)
   (@pattern_ident _ _ p) (@eta_pattern_ident_cps_gen _ _ p) (@split_types _ _ p)).

(* args_types replaces the argument p of arg_types_of with _ *)
Notation arg_types := (@arg_types_of _).

Check arg_types.
(* Before: args_types does not inherit the 8th maximal implicit of ident.arg_types *)
(* With the PR: args_types does inherit the 8th maximal implicit of ident.arg_types *)
```
A simplified form is:
```
Axiom f : forall (n:nat) {m}, m=n -> True.
Notation g n := (@f n).
Notation h := (@g 0).
Check h.
(* Before: @g 0 : forall m, (0 = m) -> True *)
(* With the PR: g 0 : (0 = ?m) -> True *)
```
What to do? Your use case does not seem unlegitimate. If we want to preserve it, we can change the semantics of `@g 0` when `g` is an abbreviation so that the deactivation role of `@` is taken into account after the instanciation of `g` by `0` rather than before (since after all, `@` has otherwise no effect on abbreviations). (*)

As a comparison:
```
Axiom f : nat -> forall {b}, b=true -> True.
Notation g := (@f 0).
Notation h := (@g).
Check h.
(* Before: @g : forall m, (0 = m) -> True *)
(* With the PR: g : (0 = ?m) -> True *)
```
I'm leaning towards implementing (*). What do you think?

**Comment by JasonGross** [comment]
I think (*) is a good idea, as long as the new abbreviation can still inherit implicit status.  That is, my expectation is that `Check @arg_types` will not fill in the 8th implicit, while `Check arg_types` will.  Is this (compatible with) what you're suggesting?

**Comment by herbelin** [comment]
About (*), I got confused. It would give the same result as what the PR already does.

> That is, my expectation is that `Check @arg_types` will not fill in the 8th implicit, while `Check arg_types` will. Is this (compatible with) what you're suggesting?

This is the case with the PR. As a consequence, I'll submit a rewriter PR to change `arg_types` into `@arg_types` there.

File: interp/notation_ops.ml [hunk 1]
```diff
@@ -187,7 +187,7 @@ let apply_cases_pattern ?loc (ids_disjpat,id) c =
 let glob_constr_of_notation_constr_with_binders ?loc g f ?(h=default_binder_status_fun) e nc =
   let lt x = DAst.make ?loc x in lt @@ match nc with
   | NVar id -> GVar id
-  | NApp (a,args) -> let e = h.no e in GApp (f e a, List.map (f e) args)
+  | NApp (a,args) -> let e = h.no e in DAst.get (mkGApp (f e a) (List.map (f e) args))
```
Comment (maximedenes) [inline interp/notation_ops.ml:190]: Couldn't this class of bugs be prevented by using an abstract type (with a view), and smart constructors, like we do for `constr`?

Comment (herbelin) [reply interp/notation_ops.ml:190]: Why not, but that would lead the PR much further that its intent, so not in this PR.

**Review by ejgallego — APPROVED** [review]
LGTM.

**Comment by ejgallego** [comment]
@coqbot: merge now

**Comment by Zimmi48** [comment]
Two external projects had to be fixed to allow merging this PR. Is this really something that we can backport and ship in 8.12.1 @herbelin @ejgallego?

**Comment by herbelin** [comment]
> Two external projects had to be fixed to allow merging this PR. Is this really something that we can backport and ship in 8.12.1 @herbelin @ejgallego?

It is a fix but I would say that there is no urgency. Maybe @jasongross has an opinion.

**Comment by JasonGross** [comment]
I see no strong reason to backport this

**Comment by ejgallego** [comment]
IMHO as it is a bugfix I'd rather ship in 8.12.1, which already has significant changes in notations.

**Comment by Zimmi48** [comment]
A bug fix with compatibility impact on two of the projects we test. Unless there was a strong reason, including it would be quite opposed to the usual compatibility policy at the patch-level.

**Pull request merged**